expression/builtin_cast_vec.go

```diff
@@ -1407,11 +1407,42 @@ func (b *builtinCastStringAsRealSig) vecEvalReal(input *chunk.Chunk, result *chu
 }
 
 func (b *builtinCastStringAsDecimalSig) vectorized() bool {
-	return false
+	return true
 }
 
 func (b *builtinCastStringAsDecimalSig) vecEvalDecimal(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	if IsBinaryLiteral(b.args[0]) {
+		return b.args[0].VecEvalDecimal(b.ctx, input, result)
+	}
+	n := input.NumRows()
+	buf, err := b.bufAllocator.get(types.ETString, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err = b.args[0].VecEvalString(b.ctx, input, buf); err != nil {
+		return err
+	}
+	result.ResizeDecimal(n, false)
+	result.MergeNulls(buf)
+	res := result.Decimals()
+	stmtCtx := b.ctx.GetSessionVars().StmtCtx
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
+			continue
+		}
+		dec := new(types.MyDecimal)
+		if !(b.inUnion && mysql.HasUnsignedFlag(b.tp.Flag) && dec.IsNegative()) {
+			if err := stmtCtx.HandleTruncate(dec.FromString([]byte(buf.GetString(i)))); err != nil {
+				return err
+			}
+			res[i] = *dec
+			if _, err := types.ProduceDecWithSpecifiedTp(&res[i], b.tp, stmtCtx); err != nil {
```

Comment:
This is different from the non-vectorized implementation. It should be:
```golang
des, err := types.ProduceDecWithSpecifiedTp(des, b.tp, stmtCtx)
if err != nil {
    return err
}
res[i] = *des
```

Reply:
I have changed it